pkg/controller/jwks/jwks.go

```diff
@@ -0,0 +1,132 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package jwks handles returning JSON encoded information about the
+// server's encryptionn keys. See https://tools.ietf.org/html/rfc75170
+// for information about the server.
+package jwks
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/google/exposure-notifications-server/pkg/logging"
+	"github.com/google/exposure-notifications-verification-server/pkg/cache"
+	"github.com/google/exposure-notifications-verification-server/pkg/database"
+	"github.com/google/exposure-notifications-verification-server/pkg/keyutils"
+	"github.com/google/exposure-notifications-verification-server/pkg/render"
+	"github.com/gorilla/mux"
+	"github.com/rakutentech/jwk-go/jwk"
+
+	"go.uber.org/zap"
+)
+
+// Error codes
+var errBadRealm = errors.New("bad realm")
+var errEncode = errors.New("error encoding key")
+
+// Controller holds all the pieces necessary to show the jwks encoded keys.
+type Controller struct {
+	h        *render.Renderer
+	db       *database.Database
+	keyCache *keyutils.PublicKeyCache
+	cacher   cache.Cacher
+	logger   *zap.SugaredLogger
+}
+
+// HandleIndex returns an http.Handler that handles jwks GET requests.
+func (c *Controller) HandleIndex() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		var keys []*database.SigningKey
+		if err := c.cacher.Fetch(ctx, r.URL.String(), &keys, 5*time.Minute, func() (interface{}, error) {
+			// Grab the URL path components  we need.
+			realmID := mux.Vars(r)["realm"]
+
+			// Find the realm, and the key abstractions from the DB.
+			realm, err := c.getRealm(realmID)
+			if err != nil {
+				return nil, err
+			}
+			var keys []*database.SigningKey
+			keys, err = realm.ListSigningKeys(c.db)
+			if err != nil {
+				return nil, err
+			}
+			return keys, nil
+		}); err != nil {
+			c.renderError(w, err)
+			return
+		}
+
+		// Get the keys.
+		encoded := make([]*jwk.JWK, len(keys))
+		for i, key := range keys {
+			pk, err := c.keyCache.GetPublicKey(ctx, key.KeyID, c.db.KeyManager())
+			if err != nil {
+				c.renderError(w, err)
+				return
+			}
+
+			// Encode it, and sent it off.
+			spec := jwk.NewSpec(pk)
+			spec.KeyID = key.GetKID()
+			encoded[i], err = spec.ToJWK()
+			if err != nil {
+				c.renderError(w, errEncode)
+				return
+			}
+		}
+		c.h.RenderJSON(w, http.StatusOK, encoded)
+	})
+}
+
+// getRealm finds realm given ID.
+func (c *Controller) getRealm(realmStr string) (*database.Realm, error) {
+	realmID, err := strconv.Atoi(realmStr)
+	if err != nil {
+		return nil, errBadRealm
+	}
+	var realm *database.Realm
+	realm, err = c.db.FindRealm(uint(realmID))
+	if err != nil {
+		return nil, errBadRealm
+	}
+	return realm, nil
+}
+
+// reportError renders an error.
+func (c *Controller) renderError(w http.ResponseWriter, err error) {
```

Comment:
there are existing helpers for this already

Reply:
Done.